### PR TITLE
Feature/releases/improve releases

### PR DESCRIPTION
**Pull request by bryanhuntesl**
Hey @sescobb27 - can you take a look at this an tell me what you think - I've got it to run successfully in production. 

~Just noticed ` _build/prod/rel/release_admin/bin/release_admin migrate` is failing - I'll see if I can resolve that now.~

### CODE REVIEW

**Comment by bryanhuntesl** [comment]
Interesting - looks like a telemetery error - 
```
/code/esl/ci/release_admin/ _build/prod/rel/release_admin/bin/release_admin migrate
Loading release_admin..
Starting dependencies..
Starting repo..
Runing migrations..
Running migrations for release_admin
16:43:26.552 [error] an exception was raised logging %DBConnection.LogEntry{call: :prepare_execute, connection_time: 1080000, decode_time: 1504000, params: [], pool_time: 47625000, query: %Postgrex.Query{cache: :reference, columns: nil, name: "", param_formats: nil, param_oids: nil, param_types: nil, ref: nil, result_formats: nil, result_oids: nil, result_types: nil, statement: ["CREATE TABLE ", "IF NOT EXISTS ", [34, "schema_migrations", 34], 32, 40, [[[], [[34, "version", 34], 32, "bigint", [[], []]], ", "], [34, "inserted_at", 34], 32, ["timestamp", "(0)"], [[],
[]]], [", ", "PRIMARY KEY (", [[], 34, "version", 34], ")"], 41, []], types: nil}, result: {:ok, %Postgrex.Query{cache: :reference, columns: nil, name: "", param_formats: [], param_oids: [], param_types: [], ref: #Reference<0.67535495.1402732546.27978>, result_formats: [], result_oids: nil, result_types: nil, statement: ["CREATE TABLE ", "IF NOT EXISTS ", [34, "schema_migrations", 34], 32, 40, [[[], [[34, "version", 34], 32, "bigint", [[], []]], ", "], [34, "inserted_at", 34], 32, ["timestamp", "(0)"], [[], []]], [", ", "PRIMARY KEY (", [[], 34, "version", 34], ")"], 41, []], types: {Postgrex.DefaultTypes, #Reference<0.67535495.1402863618.27936>}}, %Postgrex.Result{columns: nil, command: :create_table, connection_id: 5964, messages: [%{code: "42P07", file: "parse_utilcmd.c", line: "206", message: "relation \"schema_migrations\" already exists, skipping", routine: "transformCreateStmt", severity: "NOTICE", unknown: "NOTICE"}], num_rows: 0, rows: nil}}}: ** (ArgumentError) argument error
    (stdlib) :ets.lookup(:telemetry_handler_table, [:release_admin, :repo, :query])
    (telemetry) /code/esl/ci/release_admin/deps/telemetry/src/telemetry_handler_table.erl:47: :telemetry_handler_table.list_for_event/1
    (telemetry) /code/esl/ci/release_admin/deps/telemetry/src/telemetry.erl:107: :telemetry.execute/3
    (ecto_sql) lib/ecto/adapters/sql.ex:786: Ecto.Adapters.SQL.log/4
    (db_connection) lib/db_connection.ex:1311: DBConnection.log/5
    (postgrex) lib/postgrex.ex:182: Postgrex.query_prepare_execute/4
    (ecto_sql) lib/ecto/adapters/sql.ex:307: Ecto.Adapters.SQL.query!/4
    (elixir) lib/enum.ex:1314: Enum."-map/2-lists^map/1-0-"/2

{"init terminating in do_boot",{#{'__exception__'=>true,'__struct__'=>'Elixir.Ecto.MigrationError',message=><<77,105,103,114,97,116,105,111,110,115,32,102,97,105,108,101,100,32,116,111,32,114,117,110,32,98,101,99,97,117,115,101,32,116,104,101,32,99,111,110,110,101,99,116,105,111,110,32,112,111,111,108,32,115,105,122,101,32,105,115,32,108,101,115,115,32,116,104,97,110,32,50,46,10,10,69,99,116,111,32,114,101,113,117,105,114,101,115,32,97,32,112,111,111,108,32,115,105,122,101,32,111,102,32,97,116,32,108,101,97,115,116,32,50,32,116,111,32,115,117,112,112,111,114,116,32,99,111,110,99,117,114,114,101,110,116,32,109,105,103,114,97,116,111,114,115,46,10,87,104,101,110,32,109,105,103,114,97,116,105,111,110,115,32,114,117,110,44,32,69,99,116,111,32,117,115,101,115,32,111,110,101,32,99,111,110,110,101,99,116,105,111,110,32,116,111,32,109,97,105,110,116,97,105,110,32,97,32,108,111,99,107,32,97,110,100,10,97,110,111,116,104,101,114,32,116,111,32,114,117,110,32,109,105,103,114,97,116,105,111,110,115,46,10,10,73,102,32,121,111,117,32,97,114,101,32,114,117,110,110,105,110,103,32,109,105,103,114,97,116,105,111,110,115,32,119,105,116,104,32,77,105,120,44,32,121,111,117,32,99,97,110,32,105,110,99,114,101,97,115,101,32,116,104,101,32,110,117,109,98,101,114,10,111,102,32,99,111,110,110,101,99,116,105,111,110,115,32,118,105,97,32,116,104,101,32,112,111,111,108,32,115,105,122,101,32,111,112,116,105,111,110,58,10,10,32,32,32,32,109,105,120,32,101,99,116,111,46,109,105,103,114,97,116,101,32,45,45,112,111,111,108,45,115,105,122,101,32,50,10,10,73,102,32,121,111,117,32,97,114,101,32,114,117,110,110,105,110,103,32,116,104,101,32,69,99,116,111,46,77,105,103,114,97,116,111,114,32,112,114,111,103,114,97,109,109,97,116,105,99,97,108,108,121,44,32,121,111,117,32,99,97,110,32,99,111,110,102,105,103,117,114,101,10,116,104,101,32,112,111,111,108,32,115,105,122,101,32,118,105,97,32,121,111,117,114,32,97,112,112,108,105,99,97,116,105,111,110,32,99,111,110,102,105,103,58,10,10,32,32,32,32,99,111,110,102,105,103,32,58,109,121,95,97,112,112,44,32,82,101,112,111,44,10,32,32,32,32,32,32,46,46,46,44,10,32,32,32,32,32,32,112,111,111,108,95,115,105,122,101,58,32,50,32,35,32,97,116,32,108,101,97,115,116,10>>},[{'Elixir.Ecto.Adapters.SQL',raise_pool_size_error,0,[{file,"lib/ecto/adapters/sql.ex"},{line,734}]},{'Elixir.Ecto.Adapters.SQL',lock_for_migrations,5,[{file,"lib/ecto/adapters/sql.ex"},{line,719}]},{'Elixir.Ecto.Migrator',lock_for_migrations,3,[{file,"lib/ecto/migrator.ex"},{line,318}]},{'Elixir.ReleaseAdmin.ReleaseTasks.Common',setup,1,[{file,"lib/release_admin/release_tasks/common.ex"},{line,17}]},{init,start_em,1,[]},{init,do_boot,3,[]}]}}
init terminating in do_boot ({,[{Elixir.Ecto.Adapters.SQL,raise_pool_size_error,0,[{_},{_}]},{Elixir.Ecto.Adapters.SQL,lock_for_migrations,5,[{_},{_}]},{Elixir.Ecto.Migrator,lock_for_migrations,3,[{_}                                                                                     

Crash dump is being written to: erl_crash.dump...done
```

**Comment by bryanhuntesl** [comment]
Crash is Telemetry / Ecto 3.0 upgrade - https://github.com/elixir-ecto/ecto/issues/2793

File: lib/release_admin/services/release_poller.ex [hunk 1]
```diff
@@ -12,7 +12,7 @@ defmodule ReleaseAdmin.Services.ReleasePoller do
         :adapter
       ])
 
-    case RuntimeConfig.get_rpc_impl().start_polling_repo(repo_payload) do
+    case RPC.start_polling_repo(repo_payload) do
```
Comment (sescobb27) [inline lib/release_admin/services/release_poller.ex]: i think we were using this for injecting a custom RPC implementation for testing, but not sure. i think we should review tests and see how this affect them.

File: README.md [hunk 2]
```diff
@@ -16,29 +16,54 @@ You will need to create a new OAuth application on Github and provide it's `CLIE
...
 Having created the application, you will be given the opportunity to copy the GITHUB_CLIENT_ID, and GITHUB_CLIENT_SECRET - Github offers the option to regenerate the secret later if need be.
 
-3. Export these environmnental variables into the environment in which you will run the 'release_admin' application:
+
+## Export environmental variables
+
+Export these environmnental variables into the environment in which you will run the 'release_admin' application:
+
+```
+export GITHUB_CLIENT_ID=<GITHUB_CLIENT_ID>
```
[the remark this answers is not in the document]
Comment (sescobb27) [reply README.md]: are we back to not using the project name in the ENV vars?

File: Dockerfile [hunk 3]
```diff
@@ -46,4 +46,4 @@ RUN mix release
 ENTRYPOINT ["_build/prod/rel/release_admin/bin/release_admin"]
 
 # run phoenix in production on PORT 4000
-CMD ["foreground"]
+CMD ["help"]
```
[the remark this answers is not in the document]
Comment (bryanhuntesl) [reply Dockerfile]: I want to make it explicit that you provide args at startup and what those args are (self documenting) - in k8s the custome scripts we provide will be used by [init containers](https://kubernetes.io/docs/concepts/workloads/pods/init-containers/#what-can-init-containers-be-used-for) - specifically in order to ensure the database migrations are run perform the database migration at startup.

File: README.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (bryanhuntesl) [reply README.md]: Yep - it's going to be redundant when initializing from docker-compose or kubernetes.

File: config/config.exs [hunk 4]
```diff
@@ -29,9 +20,7 @@ config :ueberauth, Ueberauth,
   ]
 
 config :ueberauth, Ueberauth.Strategy.Github.OAuth,
-  client_id: System.get_env("RELEASE_ADMIN_GITHUB_CLIENT_ID"),
-  client_secret: System.get_env("RELEASE_ADMIN_GITHUB_CLIENT_SECRET")
+  client_id: {:system, "GITHUB_CLIENT_ID"},
```
[the remark this answers is not in the document]
Comment (bryanhuntesl) [reply config/config.exs]: As above

File: lib/release_admin/release_tasks/common.ex [hunk 5]
```diff
@@ -14,7 +14,7 @@ defmodule ReleaseAdmin.ReleaseTasks.Common do
     Enum.each(@start_apps, &Application.ensure_all_started/1)
 
     IO.puts("Starting repo..")
-    RuntimeConfig.get_repo().start_link(pool_size: 1)
+    Repo.start_link(pool_size: 2)
```
[the remark this answers is not in the document]
Comment (bryanhuntesl) [reply lib/release_admin/release_tasks/common.ex]: The migrations wont run unless the connection pool size is a [minimum of 2](https://github.com/elixir-ecto/ecto/issues/2245) - weird, I know.

File: lib/release_admin/services/release_poller.ex (hunk 1 above)
Comment (bryanhuntesl) [reply lib/release_admin/services/release_poller.ex]: The config is accessed within RPC - https://github.com/esl/release_admin/blob/feature/releases/improve-releases/lib/release_admin/services/rpc.ex#L7-L10

**Comment by bryanhuntesl** [comment]
@sescobb27 Tests are working locally - failing on travis because it can't download ubuntu packages - 
```
W: There is no public key available for the following key IDs:
6B05F25D762E3157  
E: Failed to fetch http://security.ubuntu.com/ubuntu/dists/trusty-security/main/binary-i386/Packages.gz  Hash Sum mismatch
E: Failed to fetch http://security.ubuntu.com/ubuntu/dists/trusty-security/universe/binary-i386/Packages.gz  Hash Sum mismatch
E: Some index files failed to download. They have been ignored, or old ones used instead.
The command "sudo apt-get update && sudo apt-get install postgresql-client" failed and exited with 100 during .
```
Re-triggered build - will see what happens and diagnose.

**Comment by bryanhuntesl** [comment]
Seems if you aren't explicit Travis randomly chooses a distro to build upon - I've explicity specified xenial now - last build failed because it tried to run on distro that had old Postgres without Xenial support - 
```
Generated release_admin app
** (Postgrex.Error) ERROR 42704 (undefined_object) type "jsonb" does not exist
    (ecto_sql) lib/ecto/adapters/sql.ex:620: Ecto.Adapters.SQL.raise_sql_call_error/1
    (elixir) lib/enum.ex:1314: Enum."-map/2-lists^map/1-0-"/2
    (ecto_sql) lib/ecto/adapters/sql.ex:707: Ecto.Adapters.SQL.execute_ddl/4
    (ecto_sql) lib/ecto/migration/runner.ex:301: Ecto.Migration.Runner.log_and_execute_ddl/3
    (ecto_sql) lib/ecto/migration/runner.ex:110: anonymous fn/2 in Ecto.Migration.Runner.flush/0
    (elixir) lib/enum.ex:1925: Enum."-reduce/3-lists^foldl/2-0-"/3
    (ecto_sql) lib/ecto/migration/runner.ex:108: Ecto.Migration.Runner.flush/0
    (stdlib) timer.erl:166: :timer.tc/1
The command "mix test" exited with 1.
```

**Comment by bryanhuntesl** [comment]
Ready for merge @sescobb27

File: lib/release_admin/services/release_poller.ex (hunk 1 above)
Comment (sescobb27) [reply lib/release_admin/services/release_poller.ex]: hahahah ups, forgot about that 👍

**Pull request merged**